feat: Add Integration through UI without breaking old config

sayam93 opened the pull request:
Allow users to add this integration using UI without breaking old configuration. Any notify service previously created through manual code in configuration.yaml remains functional. It gives the user choice to setup through UI if they want or continue with the old method.

While setup through UI, the user enters the Whatsapp number (without country code) and country code separately.

If the user's number is '+919876556789' where country code is '+91', then the user shall to enter 9876556789 as the WhatsApp Phone Number and +91 as the Country Code during setup. A notification service name notify.whatspie_919876556789 will be saved in such case.

When calling the notify service (like notify.whatspie_919876556789), the user has to mention the full phone number of the recipient with country code.

An updated readme is also available for users.

sayam93 committed:
feat: Add Integration through UI without breaking old config

Allow users to add this integration using UI without breaking old configuration. Any notify service previously created through manual code in configuration.yaml remains functional. It gives the user choice to setup through UI if they want or continue with the old method.

While setup through UI, the user enters the Whatsapp number (without country code) and country code separately.

If the user's number is '+919876556789' where country code is '+91', then the user shall to enter 9876556789 as the WhatsApp Phone Number and +91 as the Country Code during setup. A notification service name notify.whatspie_919876556789 will be saved in such case.

When calling the notify service (like notify.whatspie_919876556789), the user has to mention the full phone number of the recipient with country code.

An updated readme is also available for users.

sayam93 commented:
@arifwn could you kindly take a look at this?

arifwn commented:
Thank you @sayam93 , I'll review your pull request soon.

sayam93 commented:
> Thank you @sayam93 , I'll review your pull request soon.

@arifwn You're welcome. Let me know if there's any issue.

arifwn commented:
Hi @sayam93,

I found two issues in the pull requests when I tested it on homeassistant version 2024.12.3 :

The first one happen when adding device via configflow which results in `auth_error`. This is fixed by updating the `config_flow.py` file line  39, by changing `await self._test_credentials(user_input[CONF_API_TOKEN], full_number, country_code)` to `await self._test_credentials(user_input[CONF_API_TOKEN], full_number)` . For context, I have removed my previous settings from my `configuration.yaml` file before trying to setup a new device.

The second one happens when the device has been configured. Attempting to configure the device again via ui will result in http 500. The error go away by updating `async_get_options_flow` method to return `WhatsPieOptionsFlowHandler()` instead of `WhatsPieOptionsFlowHandler(config_entry)`, but then it'll cause the phone number and country code fields to be empty. For context, I have removed my previous settings from my `configuration.yaml` file (same config with the previous error).

Other than the two issues above, the update works on homeassistant version 2024.12.3. Existing config is still used if it exists, and the new device added from the ui can be used via `notify.whatsie_<phone number>` notification. This is a great update, thank you!

sayam93 commented:
> Hi @sayam93,
> 
> I found two issues in the pull requests when I tested it on homeassistant version 2024.12.3 :
> 
> The first one happen when adding device via configflow which results in `auth_error`. This is fixed by updating the `config_flow.py` file line 39, by changing `await self._test_credentials(user_input[CONF_API_TOKEN], full_number, country_code)` to `await self._test_credentials(user_input[CONF_API_TOKEN], full_number)` . For context, I have removed my previous settings from my `configuration.yaml` file before trying to setup a new device.
> 
> The second one happens when the device has been configured. Attempting to configure the device again via ui will result in http 500. The error go away by updating `async_get_options_flow` method to return `WhatsPieOptionsFlowHandler()` instead of `WhatsPieOptionsFlowHandler(config_entry)`, but then it'll cause the phone number and country code fields to be empty. For context, I have removed my previous settings from my `configuration.yaml` file (same config with the previous error).
> 
> Other than the two issues above, the update works on homeassistant version 2024.12.3. Existing config is still used if it exists, and the new device added from the ui can be used via `notify.whatsie_<phone number>` notification. This is a great update, thank you!

I'll look into the above two issue and update the PR today.

sayam93 commented:
> Hi @sayam93,
> 
> I found two issues in the pull requests when I tested it on homeassistant version 2024.12.3 :
> 
> The first one happen when adding device via configflow which results in `auth_error`. This is fixed by updating the `config_flow.py` file line 39, by changing `await self._test_credentials(user_input[CONF_API_TOKEN], full_number, country_code)` to `await self._test_credentials(user_input[CONF_API_TOKEN], full_number)` . For context, I have removed my previous settings from my `configuration.yaml` file before trying to setup a new device.
> 
> The second one happens when the device has been configured. Attempting to configure the device again via ui will result in http 500. The error go away by updating `async_get_options_flow` method to return `WhatsPieOptionsFlowHandler()` instead of `WhatsPieOptionsFlowHandler(config_entry)`, but then it'll cause the phone number and country code fields to be empty. For context, I have removed my previous settings from my `configuration.yaml` file (same config with the previous error).
> 
> Other than the two issues above, the update works on homeassistant version 2024.12.3. Existing config is still used if it exists, and the new device added from the ui can be used via `notify.whatsie_<phone number>` notification. This is a great update, thank you!

@arifwn code updated now. this should fix both the issues pointed out by you.

sayam93 commented:
<img width="587" alt="Screenshot 2024-12-18 at 8 51 50 PM" src="https://github.com/user-attachments/assets/4b6c056f-f0d8-45c7-958c-f3e63ce236d0" />
<img width="589" alt="Screenshot 2024-12-18 at 8 52 05 PM" src="https://github.com/user-attachments/assets/9e05bd2f-4c53-474d-9dd6-9b24cbdfabed" />

@arifwn Above changes made to the initial PR to fix the issues mentioned by you.

arifwn commented:
Thank you @sayam93 , I'll test it again soon.

arifwn commented:
@sayam93 the edit form is no longer crashing, but the phone number and country code field are empty:

https://github.com/user-attachments/assets/34f19f2c-bb0f-4706-b79a-8bee7583cbfc

sayam93 commented:
> @sayam93 the edit form is no longer crashing, but the phone number and country code field are empty:
> 
>  Screencast.From.2024-12-20.21-33-09.mp4

@arifwn Should the integration auto populate these fields with the existing config?

sayam93 commented:
> @sayam93 the edit form is no longer crashing, but the phone number and country code field are empty:
> 
>  Screencast.From.2024-12-20.21-33-09.mp4

@arifwn PR updated to fix this.

This is the change made in the latest version to fix the issue.

<img width="868" alt="Screenshot 2024-12-20 at 9 11 21 PM" src="https://github.com/user-attachments/assets/ba771989-1865-45ff-aa26-a8cfbdaaf068" />

sayam93 commented:
> @sayam93 the edit form is no longer crashing, but the phone number and country code field are empty:
> 
>  Screencast.From.2024-12-20.21-33-09.mp4

@arifwn I have updated the PR to fix this by adding two new constants to const.py in addition to the existing 4.

I have also imported these constants in the config_flow.py to store unsanitised value from user input while setting up the configuration entry using UI.

These values are then showed up while editing the config and when edited, correctly combined to rebuild the number to be stored in the config.

See the below screenshots for changes.

<img width="624" alt="Screenshot 2024-12-20 at 10 33 45 PM" src="https://github.com/user-attachments/assets/30e09448-753c-4ee2-a09b-8bca5c5c1cca" />
<img width="407" alt="Screenshot 2024-12-20 at 10 33 57 PM" src="https://github.com/user-attachments/assets/df33e5ab-bc9b-42a1-bc8c-1806c8fe6c43" />
<img width="893" alt="Screenshot 2024-12-20 at 10 34 47 PM" src="https://github.com/user-attachments/assets/abdbab63-6ef1-40de-9e22-174929380558" />
<img width="628" alt="Screenshot 2024-12-20 at 10 35 03 PM" src="https://github.com/user-attachments/assets/fde06d2c-e655-422e-97f6-ea6e298054ef" />
<img width="366" alt="Screenshot 2024-12-20 at 10 35 28 PM" src="https://github.com/user-attachments/assets/5ac388b5-6e1d-48dd-9b84-70d4954e017b" />

sayam93 commented:
@arifwn Let me know how the updated PR works for you. I am unable to test the latest changes myself as my trial has expired.

sayam93 commented:
@arifwn Let me know how the updated PR works for you. I am unable to test the latest changes for myself since my trial subscription expired.

arifwn commented:
Thank you @sayam93 , I'll do some more testing this weekend.

sayam93 commented:
> Thank you @sayam93 , I'll do some more testing this weekend.

@arifwn Do let me know if there are any issues.

sayam93 commented:
> Thank you @sayam93 , I'll do some more testing this weekend.

@arifwn Do let me know if there are any issues.

arifwn commented:
@sayam93 featurewise, I found no issue, but the manifest check below produce some error:

![image](https://github.com/user-attachments/assets/14f6b224-8d6f-4093-a486-4bff8a9615fb)

https://github.com/arifwn/homeassistant-whatspie-integration/actions/runs/12435355722

sayam93 committed:
Fix HACS Error

sayam93 commented:
> @sayam93 featurewise, I found no issue, but the manifest check below produce some error:
> 
> ![image](https://private-user-images.githubusercontent.com/1146705/399686829-14f6b224-8d6f-4093-a486-4bff8a9615fb.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzU4MzcwMzQsIm5iZiI6MTczNTgzNjczNCwicGF0aCI6Ii8xMTQ2NzA1LzM5OTY4NjgyOS0xNGY2YjIyNC04ZDZmLTQwOTMtYTQ4Ni00YmZmOGE5NjE1ZmIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDEwMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAxMDJUMTY1MjE0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NWM3MjQyMzZjODI5ZjliZTBmMGE4NTZjODlkZDUwNzM3NDNkYzdiNDUzODcyY2M0NjM0MjdkY2FjMWEwNTc0YiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.V-mIDInaO4ZOoIOJl3DUG6ggbk0sKW0Ci2pADA7peBA)
> 
> https://github.com/arifwn/homeassistant-whatspie-integration/actions/runs/12435355722

I think the latest commit should fix all these errors.

sayam93 commented:
> @sayam93 featurewise, I found no issue, but the manifest check below produce some error:
> 
> ![image](https://private-user-images.githubusercontent.com/1146705/399686829-14f6b224-8d6f-4093-a486-4bff8a9615fb.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzU4MzcwMzQsIm5iZiI6MTczNTgzNjczNCwicGF0aCI6Ii8xMTQ2NzA1LzM5OTY4NjgyOS0xNGY2YjIyNC04ZDZmLTQwOTMtYTQ4Ni00YmZmOGE5NjE1ZmIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDEwMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAxMDJUMTY1MjE0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NWM3MjQyMzZjODI5ZjliZTBmMGE4NTZjODlkZDUwNzM3NDNkYzdiNDUzODcyY2M0NjM0MjdkY2FjMWEwNTc0YiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.V-mIDInaO4ZOoIOJl3DUG6ggbk0sKW0Ci2pADA7peBA)
> 
> https://github.com/arifwn/homeassistant-whatspie-integration/actions/runs/12435355722

@arifwn The latest commit should fix all these errors.

arifwn commented:
Thank you @sayam93! There is some minor manifest issue left, but those are minor issues and I can fix that when I merge this pr soon.
